### ISSUE TITLE
QUESTION: independent tasks to be run without ordering

### ISSUE THREAD

**Issue by RiccardoRossi** [cpp-taskflow/cpp-taskflow#74]
This is even less than a feature request, and more of a conceptual question on tasking.

let's imagine that i have the tasks

A B C D E F G

because of their structure the tasks can be run **in any order**. HOWEVER a task can be running only when its neighbour tasks are not being run (however it does not matter if the neighbouring tasks were actually executed or not).
To make an example, i could run B only under the guarantee that neither A or C are run while B is running.

the question is:
how can i prescribe such logical dependency without implying an order between the tasks?

thank you in advance for any hint on this

**Comment by RiccardoRossi** [comment]
Dear @clin99, sorry for very long delay in answering. I have been extremely busy and then recently travelling.

I am convinced that the implementation you propose **would actually work**, nevertheless if i am not wrong it implies an ordering in the execution. let me try to vocalize this better:

if you start with 
A B C D E F G

your independent set would be
A   C   E   G

if my understanding is correct, you will never execute for example D until you executed both C and E, while really you would be able to execute D under the only constraint that C and E are not executed in the same moment.

I think this is an example in which a DAG is not the correct way to express dependencies (nevertheless i don't have a clue on how else this should be expressed)

In any case  thx a lot for the feedback!

**Comment by clin99** [comment]
Hi @RiccardoRossi , appreciate your feedback!.  Yes, your execution order is correct. 

Let me rephrase the example:

A-B-C-D-E-F-G     ('-' means the two tasks are neighbors (not dependencies)) 

If I select the independent set greedily following the alphabetical order, (A C E G) will be executed first and then (B D F). In this case, D will start execution after both C and E finish. 

However, it's also correct to execute (B D F) first and then (A C E G). This all depends on how you select the independent set!   Do you wish to impose an execution order on the tasks along with the neighbor constraint?

Thanks!

**Comment by RiccardoRossi** [comment]
@clin99 thank you for your feedback. however let me iterate:
the entire point of this "issue" is in NOT prescribing the order before runtimr. i would call it an academic curiosity 😅

i mean, we can definitely construct an execution tree as you propose (or in other imaginative ways), nevertheless if we do not do it wisely (for example if the tasks take very different time to execute) thry will run in parallel but suboptimally.

just think at the case Admittedly  in which A and D arr way slower than all the others.

as i understand D will be run after A C E G are all finished, while we could run it before A finishes.

**Comment by clin99** [comment]
Hi @RiccardoRossi , thanks for pointing out this. I think my explanation about the example is a bit misleading. 

In that example, I put (A C E G) and (B D F) in two groups in order to differentiate the 
independent sets, while this doesn't mean (B D F) must be executed together after (A C E G) all finish.
In fact, D will be dispatched to execution immediately after C and E finish and D does not need to wait for A and G (A and G are not neighbors of D!).  Also in the pseudo code, you can see a task dispatches it's neighbor to execution as soon as the neighbor becomes ready (it has no running neighbors). I hope this addresses your concern and sorry for causing the confusion. 

Please let me know if this does answer your question.

Thanks!

**Comment by RiccardoRossi** [comment]
oh i see!.

i thought that the whole batch had to finish. than your solution is completely satisfactory to me! thx!! 

I'll close this
